types.go

```diff
@@ -121,6 +122,18 @@ type functionInfo struct {
 	statuser         Statuser
 }
 
+// TODO(btracey): Think about making this an exported function when the
+// constraint interface is designed.
+func (f functionInfo) satisfies(method Method) error {
+	gradOk := f.IsGradient || f.IsFunctionGradient
+	if method.Needs().Gradient && !gradOk {
+		return errors.New("function does not implement Gradient or FunctionGradient")
```

Comment:
@kortschak do errors get "optimize: " prefixes?

This error needs to be a bit better. The error is that the Method needs the gradient and that the Function can not supply it.

"Method needs gradient information but Function does not implement Gradient or FunctionGradient"